File: Easy/7. Reverse Integer/official solution (1).cpp
// 7. Reverse Integer

// Runtime: 4 ms, faster than 48.54% of C++ online submissions for Reverse Integer.

// Memory Usage: 6.1 MB, less than 100.00% of C++ online submissions for Reverse Integer.


class Solution {
public:
    int reverse(int x) {
        int rev = 0;
        while (x != 0) {
            int pop = x % 10;
            x /= 10;
            if (rev > INT_MAX/10 || (rev == INT_MAX / 10 && pop > 7)) return 0;
            if (rev < INT_MIN/10 || (rev == INT_MIN / 10 && pop < -8)) return 0;
            rev = rev * 10 + pop;
        }
        return rev;
    }
};

File: Easy/13. Roman to Integer/official solution (1).py
# 13. Roman to Integer

# Runtime: 40 ms, faster than 90.29% of Python3 online submissions for Roman to Integer.
# Memory Usage: 13.8 MB, less than 5.38% of Python3 online submissions for Roman to Integer.


class Solution:
    def romanToInt(self, s: str) -> int:
        if len(s) == 0:
            return 0

        dict = {'I':1,"V":5,"X":10,"L":50,"C":100,"D":500,"M":1000}
        prev = dict[s[0]]
        sum  = prev
        for i in range(1, len(s)):
            curr = dict[s[i]]
            sum += curr
            if curr > prev:
                sum -= prev * 2
            prev = curr
        return sum